### PR TITLE
Add org.freedesktop.Sdk.Extension.wxwidgets

### PR DESCRIPTION
**Pull request by tinywrkb**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link: https://github.com/wxWidgets/wxWidgets/issues/22370**
- [ ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

This SDK extension packages wxWidgets, WebKitGTK, and their dependencies.  
I'm not developing wxWidgets apps, but a while back I took over the maintenance of Codeblocks IDE (I was nostalgic), which its maintainer and original submitter was MIA.  
As a wxWidgets app, an important feature of Codeblocks is its support for wxWidgets development, and the previous Flatpak maintainer did not take this into consideration.  

When I started maintaining the app, I quickly switched the runtime to Gnome due to the availability WebKitGTK, and bundled wxWidgets' development files (headers, pkgconfig, ...) to have some sort of wxWidgets development environment, but I don't think this is the way forward.  
Instead, I'm proposing adding a wxWidgets Freedesktop SDK extension which could work with any IDE.

Relevant Codeblocks app changes flathub/org.codeblocks.codeblocks#21

~~This is still WIP, so it's marked as draft.~~

### CODE REVIEW

**Comment by tinywrkb** [comment]
bot, build org.freedesktop.Sdk.Extension.wxwidgets

**Comment by tinywrkb** [comment]
Per wxWidgets' developer recommendation, I switch to wxWidgets 3.1.x development releases, and will switch back to stable after the 3.2.0 release will be out.

What's blocking this now is mainly doing more testing.  
I might revise the deployment script, `install.sh`, though that much lower priority, and might be done post merge.

**Comment by tinywrkb** [comment]
bot, build org.freedesktop.Sdk.Extension.wxwidgets

File: org.freedesktop.Sdk.Extension.wxwidgets.metainfo.xml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="runtime">
+  <id>org.freedesktop.Sdk.Extension.wxwidgets</id>
+  <metadata_license>CC0-1.0</metadata_license>
+  <project_license>wxWindows</project_license>
```
Comment (hfiguiere) [inline org.freedesktop.Sdk.Extension.wxwidgets.metainfo.xml:5]: This should do it
```suggestion
  <project_license>LGPL-2.1-or-later WITH WxWindows-exception-3.1</project_license>
```

(it's a valid SPDX identifier)

File: org.freedesktop.Sdk.Extension.wxwidgets.yaml [hunk 2]
```diff
@@ -0,0 +1,209 @@
+id: org.freedesktop.Sdk.Extension.wxwidgets
```
Comment (hfiguiere) [inline org.freedesktop.Sdk.Extension.wxwidgets.yaml:1]: I would rather take the BaseApp route instead.

Comment (tinywrkb) [reply org.freedesktop.Sdk.Extension.wxwidgets.yaml:1]: As written in the 1st post, this is designed for wxWidgets development with Codeblocks IDE or any other IDEs.  
Building WebKitGTK without ccache hits takes a substantial amount of time, and I would not bother with it if I didn't want an SDK extension.  
Even if there was a base application, I would still submit this extension.

Comment (hfiguiere) [reply org.freedesktop.Sdk.Extension.wxwidgets.yaml:1]: ah ok, nevermind then.

File: org.freedesktop.Sdk.Extension.wxwidgets.metainfo.xml (hunk 1 above)
Comment (tinywrkb) [reply org.freedesktop.Sdk.Extension.wxwidgets.metainfo.xml]: Fixed in 10e425720ae72039fb551a46c0d369a6fe2e3dec

**Comment by tinywrkb** [comment]
bot, build org.freedesktop.Sdk.Extension.wxwidgets

**Comment by tinywrkb** [comment]
bot, build org.freedesktop.Sdk.Extension.wxwidgets

**Comment by tinywrkb** [comment]
bot, build org.freedesktop.Sdk.Extension.wxwidgets

**Comment by tinywrkb** [comment]
This should be ready now.  
I went through all the code samples of wxWidgets, and they can be built and run correctly as far as I can tell.  
Also built and run with this extension the Flathub app `net.sourceforge.GrandOrgue`, and Codeblock (svn trunk version).

**Comment by tinywrkb** [comment]
Switching to draft to evaluate building WebKitGTK with full debugging symbols.

**Comment by 4c0n** [comment]
Great! Making this available as a SDK extension makes it flexible and re-usable. :smile:
I'm wondering if it would be a good idea to have WebkitGTK as a separate extension as there seem to be some other packages that currently have the `--disable-webkit` or `--without-webkit` flags in their builds, maybe because it's too time consuming to include it, but of course I can't be sure about that.

**Comment by tinywrkb** [comment]
@4c0n 

This is already being handled by the `install.sh` script, and the next push here will have examples.  
I'm currently rewriting the script, as it has a few issues, and I want it to fail if new shared libs or locales were detected to limit the maintenance burden of having to keep track of these changes.  

Next push here will have updated `install.sh`, examples, and will test full debug symbols for WebKitGTK.

**Comment by tinywrkb** [comment]
bot, build org.freedesktop.Sdk.Extension.wxwidgets

**Comment by tinywrkb** [comment]
The contribution of WebKitGTK's debug symbols to the size of this extension:

- `-g0` (no debug symbols): extension total size is ~400MB.
- `-g1`: extension total size is ~1GB
- `-g` (flatpak-builder's default): extension total size is ~4.5GB. Almost an extra 30 minutes to build the extension.  
  Also, creating `.Debug` extension for applications becomes a very costly operation.

I'm in favor of switching back setting `-g1`.

**Comment by tinywrkb** [comment]
bot, build org.freedesktop.Sdk.Extension.wxwidgets

**Comment by tinywrkb** [comment]
> Hence the reason for wondering if maybe it would be better to place it in a separate sdk extension.

No, there's no way to keep these extensions in sync and not break ABI without manual intervention of a Flathub admin keeping ostree refs from being published until both extensions are rebuilt.  
We will be going to need a separate Flatpak branch for each WebKitGTK release that breaks ABI, and after it's publish to update wxWidgets to use the new WebKitGTK branch.  
I asked similar question about handling cross-Flatpak-ref ABI dependency [here](https://github.com/flathub/flathub/issues/2533).

**Comment by tinywrkb** [comment]
Last push just renames the examples.

**Comment by tinywrkb** [comment]
> > Hence the reason for wondering if maybe it would be better to place it in a separate sdk extension.
> 
> No, there's no way to keep these extensions in sync and not break ABI without manual intervention of a Flathub admin keeping ostree refs from being published until both extensions are rebuilt. We will be going to need a separate Flatpak branch for each WebKitGTK release that breaks ABI, and after it's published to update wxWidgets to use the new WebKitGTK branch. I asked similar question about handling cross-Flatpak-ref ABI dependency [here](https://github.com/flathub/flathub/issues/2533).

After considering this again, I see that WebKitGTK has semi-annual schedule for major releases, so maybe if we're not strict as the Freedesktop runtime about adding new symbols, and don't mind stating that we can't promise that downgrading the WebKitGTK extension won't break ABI of code linked against the latest WebKitGTK build, then I guess a WebKitGTK extension is achievable without too much trouble.  
We will need 2 branches for each runtime version, so a total of 4 maintained branches, and maybe retire some of them earlier than the runtime.  
I'm not sure yet what the best way to wire-up ABI checks of minor releases, or should we just trust upstream.

**Comment by tinywrkb** [comment]
> > > Hence the reason for wondering if maybe it would be better to place it in a separate sdk extension.
> > 
> > 
> > No, there's no way to keep these extensions in sync and not break ABI without manual intervention of a Flathub admin keeping ostree refs from being published until both extensions are rebuilt. We will be going to need a separate Flatpak branch for each WebKitGTK release that breaks ABI, and after it's published to update wxWidgets to use the new WebKitGTK branch. I asked similar question about handling cross-Flatpak-ref ABI dependency [here](https://github.com/flathub/flathub/issues/2533).
> 
> After considering this again, I see that WebKitGTK has semi-annual schedule for major releases, so maybe if we're not strict as the Freedesktop runtime about adding new symbols, and don't mind stating that we can't promise that downgrading the WebKitGTK extension won't break ABI of code linked against the latest WebKitGTK build, then I guess a WebKitGTK extension is achievable without too much trouble. We will need 2 branches for each runtime version, so a total of 4 maintained branches, and maybe retire some of them earlier than the runtime. I'm not sure yet what the best way to wire-up ABI checks of minor releases, or should we just trust upstream.

I take this back. I've been criticized for doing something similar in an extension. Avoiding ABI breaking by setting a branch version.  
I'm not going to bother with this when no official Flathub guidelines document how to avoid ABI breakage between different Flatpak refs/packages.  
This is a Flatpak limitation that should be fixed, and I won't try to work around.

I have no further changes planned to this PR, and it's ready to be merged.

**Comment by 4c0n** [comment]
> > > > Hence the reason for wondering if maybe it would be better to place it in a separate sdk extension.
> > > 
> > > 
> > > No, there's no way to keep these extensions in sync and not break ABI without manual intervention of a Flathub admin keeping ostree refs from being published until both extensions are rebuilt. We will be going to need a separate Flatpak branch for each WebKitGTK release that breaks ABI, and after it's published to update wxWidgets to use the new WebKitGTK branch. I asked similar question about handling cross-Flatpak-ref ABI dependency [here](https://github.com/flathub/flathub/issues/2533).
> > 
> > After considering this again, I see that WebKitGTK has semi-annual schedule for major releases, so maybe if we're not strict as the Freedesktop runtime about adding new symbols, and don't mind stating that we can't promise that downgrading the WebKitGTK extension won't break ABI of code linked against the latest WebKitGTK build, then I guess a WebKitGTK extension is achievable without too much trouble. We will need 2 branches for each runtime version, so a total of 4 maintained branches, and maybe retire some of them earlier than the runtime. I'm not sure yet what the best way to wire-up ABI checks of minor releases, or should we just trust upstream.
> 
> I take this back. I've been criticized for doing something similar in an extension. Avoiding ABI breaking by setting a branch version.  
> I'm not going to bother with this when no official Flathub guidelines document how to avoid ABI breakage between different Flatpak refs/packages.  
> This is a Flatpak limitation that should be fixed, and I won't try to work around.
> 
> I have no further changes planned to this PR, and it's ready to be merged.

Totally agree that it would be great to have more guidelines and information on this subject.
Too bad communication seems difficult when it comes to matters like this, I recently tried proposing for more guidelines on the discourse forum, but also did not get any useful response so far.
I truly hope communication will improve for Flathub and its community.

**Comment by tinywrkb** [comment]
bot, build org.freedesktop.Sdk.Extension.wxwidgets

**Comment by tinywrkb** [comment]
bot, build org.freedesktop.Sdk.Extension.wxwidgets

**Comment by tinywrkb** [comment]
bot, build org.freedesktop.Sdk.Extension.wxwidgets

**Pull request closed without merging**